pkg/apis/clusterregistry/types.go

```diff
@@ -32,43 +32,64 @@ type Cluster struct {
 	// +optional
 	metav1.ObjectMeta
 
-	// State is the state of the cluster. This is not a specification, and is not
-	// meant to be used by actively-reconciling controllers; it is also not
-	// a status, as it contains fields that do not necessarily describe the
-	// status of the cluster, and is not necessarily updated by an active
-	// controller.
+	// Spec is the specification of the cluster. This may or may not be
+	// reconciled by an active controller.
 	// +optional
-	State ClusterState
+	Spec ClusterSpec
+
+	// Status is the status of the cluster. It is optional, and can be left nil
+	// to imply that the cluster status is not being reported.
+	// +optional
+	Status *ClusterStatus
 }
 
-// ClusterState contains the state of a cluster.
-type ClusterState struct {
+// ClusterSpec contains the spec of a cluster.
+type ClusterSpec struct {
 
-	// KubeAPIServer represents the API server for this cluster.
+	// KubernetesAPIEndpoints represents the endpoints of the API server for this
+	// cluster.
 	// +optional
-	KubeAPIServer KubeAPIServer
+	KubernetesAPIEndpoints KubernetesAPIEndpoints
 
 	// AuthInfo contains public information that can be used to authenticate
-	// to and authorize with this cluster.
+	// to and authorize with this cluster. It is not meant to store private
+	// information (e.g., tokens or client certificates) and the cluster registry
+	// implementations are not expected to provide hardened storage for
+	// secrets.
 	// +optional
 	AuthInfo AuthInfo
 
 	// CloudProvider contains information about the cloud provider this cluster
 	// is running on.
 	// +optional
-	CloudProvider CloudProvider
+	CloudProvider *CloudProvider
+}
+
+// ClusterStatus contains the status of a cluster.
+type ClusterStatus struct {
+	// TODO https://github.com/kubernetes/cluster-registry/issues/28
 }
 
-type URL string
+// ServerAddressByClientCIDR helps the client to determine the server address that they should use, depending on the clientCIDR that they match.
+type ServerAddressByClientCIDR struct {
```

Comment:
small nit: can we define this type below the `KubernetesAPIEndpoints` type?

Reply:
Sure, done.